node/runtime/pangolin/src/pallets/bridge_dispatch.rs

```diff
@@ -3,19 +3,80 @@ pub use pallet_bridge_dispatch::{
 };
 
 // --- paritytech ---
-use frame_support::traits::Everything;
+use frame_support::traits::OriginTrait;
+use sp_runtime::transaction_validity::{InvalidTransaction, TransactionValidityError};
 // --- darwinia-network ---
 use crate::*;
+use bp_message_dispatch::{CallValidate, IntoDispatchOrigin as IntoDispatchOriginT};
 use bp_messages::{LaneId, MessageNonce};
+use darwinia_ethereum::{RawOrigin, Transaction};
+use darwinia_evm::AccountBasic;
+use darwinia_support::evm::{DeriveEthereumAddress, DeriveSubstrateAddress};
 use pallet_bridge_dispatch::Config;
 
+pub struct CallValidator;
+impl CallValidate<bp_pangolin::AccountId, Origin, Call> for CallValidator {
+	fn pre_dispatch(
+		relayer_account: &bp_pangolin::AccountId,
+		origin: &Origin,
+		call: &Call,
+	) -> Result<(), TransactionValidityError> {
+		match call {
+			// Note: Only supprt Ethereum::transact(LegacyTransaction)
+			Call::Ethereum(darwinia_ethereum::Call::transact { transaction: tx }) => {
+				match origin.caller() {
+					OriginCaller::Ethereum(RawOrigin::EthereumTransaction(id)) => {
+						match tx {
+							Transaction::Legacy(t) => {
+								let fee = t.gas_limit.saturating_mul(t.gas_limit);
+								let total_payment = fee.saturating_add(t.value);
+
+								let derived_substrate_address = <Runtime as darwinia_evm::Config>::IntoAccountId::derive_substrate_address(*id);
+								if <Runtime as darwinia_evm::Config>::RingAccountBasic::account_balance(relayer_account) >= total_payment {
```

Comment:
Smart chain and substrate balance precision conversion.

10 ** 18 -> 10 **9

Reply:
```rs
fn account_balance(account_id: &T::AccountId) -> U256 {
	// Get main balance from Currency.
	let main_balance = C::free_balance(&account_id).saturated_into::<u128>();
	// Get remaining balance from Dvm.
	let remaining_balance = RB::remaining_balance(&account_id).saturated_into::<u128>();
	// final_balance = balance * 10^9 + remaining_balance.
	decimal_convert(main_balance, Some(remaining_balance))
}
```

`account_balance` has done this conversion internally.